Add reference images for KMonad Lite default and QWERTY layouts

UncleCJ opened the pull request:
I completely understand if one doesn't want to bother with all of these variations, but at least illustrations of the "lite" layout also in QWERTY variant (using KMonad) would have attracted me to the project even quicker.

Screenshots are made manually from [KLE](http://www.keyboard-layout-editor.com) at 150%, because the download options are (poorly) experimental

This is in reference to #71 (although I noticed #72 only now...) and I would love to tidy the JSON (how does KLE classify types of legends?) but no rush - use (and modify) these only if they make sense in a general plan. There are a couple more options in [UncleCJ:miryoku:development](https://github.com/manna-harbour/miryoku/compare/master...UncleCJ:miryoku:development)

manna-harbour commented:
Great idea, thanks! That'll be a nice addition!  Did you edit the Layer Reference diagram and re-arrange the keys?

The lite mapping is intended to be used just for alphas, mods, and Nav, and by using the rest of the keys as usual. (Although with the default layers option (non-flip) the thumb keys are also mapped to their corresponding Base tap keys, so they can also be used for layers as an alternative to the rest of the keys, but it might get awkward for some, especially if mods are also involved.)

With that in mind, it might be nice to un-ghost the rest of the keys and add their default legends? That would require choosing a particular layout though, so might need some consideration first.

UncleCJ commented:
Hmmm, let's see what I've been thinking... yes, I modified [data/layers/miryoku-kle-reference](/manna-harbour/miryoku/blob/master/data/layers/miryoku-kle-reference.png) to the shape of the "[Default 60%](https://github.com/ijprest/keyboard-layout-editor/blob/master/layouts.json#L28)" (and ISO variant) and as suggested in #71, replaced the base layer legends as one would for QWERTY.

However, I haven't wrapped my head entirely around [babel](/manna-harbour/miryoku_babel) yet (though it looks awesome!), so don't know what you've been doing in #72 - and looking at these raw JSON, particularly compared to the KLE samples, there may be a lot of cruft in there (I certainly don't know yet whether I maintained any order). It's what I refer to in my opening caveat, it could be a _bad idea_ to introduce anything which would be manually maintained (hmm, except for attracting nitpickers like me? :-D).

Default legends - sure! I don't think EurKey or US Intl QWERTY would be too presumptuous (yes, I too looked for AltGr as in #53). Something like my personal [UncleCJ:miryoku:development/data/layers/miryoku-kle-reference-lite-qwerty-vi-iso-apple.png](/UncleCJ/miryoku/blob/development/data/layers/miryoku-kle-reference-lite-qwerty-vi-iso-apple.png)? Digressing for a moment, but dipping my toe into something this esoteric, basically KMonad, I feel inevitably irreverent. Though certainly Miryoku Colemak-DH on QMK is something to be desired, I'm simply not there yet. So, I appreciate your insight into how Miryoku (Lite, QWERTY) is intended to be utilized, and will continue improving my setup... and yeah, am glad you seem to agree it could be a useful lowering of the barrier to entry (my first attempt, I simply didn't notice the [default, non-lite mapping was one row up](/manna-harbour/miryoku/master/data/mapping/miryoku-kle-mapping-60_ansi.png)).

With all that said, I suggest not to be in a hurry to merge this, I'll be happy to bake this a bit further. 

Slightly off topic - but as my first QMK-capable Corne is being shipped from Poland to Sweden (any thoughts on how to utilize the sixth columns?), my substitute Windows 11 KMonad setup really does not feel "production ready" - shutting KMonad down almost freeze my system, sends a bunch of 'ccccccccccc' everywhere and I probably manage to occasionally get stuck in some layers... that project being what it is, besides `--log-level debug`, do you have any advice on how to run it - automatically, user or admin level, etc.? I don't really mind doing a bit of work, difficult things are rewarding, but it would be sad if I didn't manage to stabilize a KMonad setup...

manna-harbour commented:
> Slightly off topic - but as my first QMK-capable Corne is being shipped from Poland to Sweden (any thoughts on how to utilize the sixth columns?), 

See the first paragraph here: https://github.com/manna-harbour/miryoku/discussions/98#discussioncomment-3031766. For my first crkbd, I removed the outer pinkie columns with a hacksaw!

> my substitute Windows 11 KMonad setup really does not feel "production ready" - shutting KMonad down almost freeze my system, sends a bunch of 'ccccccccccc' everywhere and I probably manage to occasionally get stuck in some layers... that project being what it is, besides --log-level debug, do you have any advice on how to run it - automatically, user or admin level, etc.? I don't really mind doing a bit of work, difficult things are rewarding, but it would be sad if I didn't manage to stabilize a KMonad setup...

I can't really help with windows, but perhaps there are more and less graceful methods for exiting. Try searching and asking on the KMonad discord and issues, I'm sure someone must have had the same problem.

Getting stuck in layers is probably the layer lock though. See https://github.com/manna-harbour/miryoku/tree/master/docs/reference#additional-features and https://github.com/manna-harbour/miryoku/discussions/93.

UncleCJ commented:
Thanks for the various advice!

Again, my context has developed - as I built my first crkbd and _1)_ miryoku_qmk didn't seem to easily support the OLED displays or RGB backlights (or Via, but again, that has its own issues) _2)_ I realized I wanted Miryoku variations not supported as options (let's see, numpad layout, F-keys and just generally additions which felt logical to me) _3)_ I certainly need the `TAPPING_TERM` increased and other options to get along with mod tapping (those seem to be suppressed if I try to use the Via keymap). Not strictly Miryoku concerns, but [my adaptions are available](/UncleCJ/qmk_firmware/commit/b2309cd8a82418?diff=unified). With that said, now that I've seen miryoku_qmk and miryoku_babel within the greater QMK tree, I'm quite awestruck! 😄 It's really quite the effort to attempt unifying enabling of great keyboard layouts!

Thus, the KLE illustration is for me personally slightly less relevant now... I did tinker some more with it, discovering I _can_ decrease redundancy in the profile as well as align it with the babel row-wise description... except KLE will helpfully transpose it and mess up my tedious effort... So will see whether I find any way forward, in that case of course I'd get in touch. Thanks for the `callum-oakley/keymap` reference! Indeed it seems a bit limited compared to KLE, yeah... we'll see 🙂

UncleCJ commented:
I redid the ANSI Lite layouts according to the comments and hope you agree they look way nicer now. As mentioned before, and even more so once the purpose of Lite variations was clarified, I think it is a really great entry level usage of Miryoku worth illustrating. In fact, I'll try again to get it going on my machines so I can still use mod tapping when my custom keyboard is not available. Thanks!

UncleCJ commented:
@manna-harbour hi! I hope you’re not feeling pestered by these comments or suggestions - I fidget with them for my own benefit, because it’s enjoyable and with a slight hope it may be useful for others.

With that in mind I wanted to write some general comments and questions. Please pardon I won’t be cross referencing right now.

1. Regarding adding images of Lite layouts, QWERTY etc and perhaps attempting a nicer color theme (I am trying out imitating "hysp") - the more I look, the more I realize any contributions need to fit into the existing documentation and usage of illustrations. For instance there are both `layers` and `mapping` folders and so far I have focused on the previous… and even though it’s the Lite mapping, I think it is more of a layer illustration… Also acting on any comments obviously pile up quickly - I did the un-ghosted lite layout and in general… do give me any feedback if you feel there is a chance I can effectively contribute, otherwise don’t fret it, I am fully aware of my… limited awareness :-D
2. It’s an obvious limitation not to have a continuous generation of the KLE diagrams… I noted some potential errors such as broken logo urls, lack of `AltGr` modifiers and the tab glyph may be funky… I may submit some trivial merge requests on things like that, assuming they are of any use
3. Contingent on the continued use of KLE… My merge request on transposing the JSON makes sense to me, the SVG render you mentioned feels too limited in comparison, though the most future proof path certainly sounds like visualizing ergogen… I saw some Miryoku QMK report (?), but one wants to stay on the “source” side rather than the generated result? Do let me know if there’s anything regarding the new customization documentation, babel or similar which you think may be promising, and I’d be happy to pivot my efforts. Continuing on KLE a bit, regarding the “transposing“, as I was reading about Kyria and other custom keyboards, it occurred that the thumb clusters logically are columns despite being quite horizontal. I’ve taken to your language of locating keys per finger, and in that sense, KLE being _row_ oriented become a severe design deficiency which makes no sense. Most, if not all of your space in babel and implementations in QMK and KMonad are also row-oriented though… though I expect ergogen to be flexible. Do you have any thoughts on that?
4. Minor thoughts… but my Crkbd has 1.5U central thumb keys and it felt a bit awkward to consider the middle ones “primary” and previously they had been Space bar and Enter, respectively. I can’t manage to author a macro for it yet but have a personal mod which swap primary and secondary, and then swap back Enter and Backspace. Do you have a strong opinion whether I may regret that, or do you even feel like adopting it? Furthermore the apostrophe and semicolon swap tripped me up (Club Colemak emulation broke, with the positive side effect that I rebased on your development branch and got ColemakDH extras working beautifully instead), I’m considering expanding/altering QWERTY with the Nordic layout (as I may convert also my partner to the crkbd team…).
5. Almost forgot, what do you think of “autoshift“ keys? I read of someone going all the way down to 30 keys, and it seemed to make a lot of sense to have several of the symbols autoshifted as on the number row (though mapped like a num pad)

All in all, I even got KMonad working decently now and with each step appreciate the logic of Miryoku, so would be quite happy both to “paint within its lines” and to contribute to it in any way I can. Thank you so much!

manna-harbour commented:
For customisation in general see https://github.com/manna-harbour/miryoku/discussions/85.

> _1)_ miryoku_qmk didn't seem to easily support the OLED displays or RGB backlights (or Via, but again, that has its own issues) 

For OLED and RGB see the `Custom Hardware Support` section in the above docs. I have something else for adding hardware support more easily, but it needs to be rewritten: https://github.com/manna-harbour/qmk_firmware/issues/40. It also needs OLED and RGB to be added from the previous version: https://github.com/manna-harbour/qmk_firmware/tree/crkbd/keyboards/crkbd/keymaps/manna-harbour.

> _2)_ I realized I wanted Miryoku variations not supported as options (let's see, numpad layout, F-keys and just generally additions which felt logical to me) 

To substitute custom layers, see the `custom_config.h` section in the above docs.

> _3)_ I certainly need the `TAPPING_TERM` increased and other options to get along with mod tapping (those seem to be suppressed if I try to use the Via keymap). 

`TAPPING_TERM` definitely needs to be easily customisable. See https://github.com/manna-harbour/miryoku/issues/94 and https://github.com/manna-harbour/miryoku/issues/33.

Via / Vial / Remap support is a whole other thing which I haven't even begun looking into...

> Not strictly Miryoku concerns, but [my adaptions are available](/UncleCJ/qmk_firmware/commit/b2309cd8a82418?diff=unified). With that said, now that I've seen miryoku_qmk and miryoku_babel within the greater QMK tree, I'm quite awestruck! smile It's really quite the effort to attempt unifying enabling of great keyboard layouts!

Nice! Using `custom_config.h` means you can customise but still pull in Miryoku updates, and without conflicts, and still be able to build for all supported keyboards. But at some point it probably makes more sense to just make your own keymap from scratch!

> Thus, the KLE illustration is for me personally slightly less relevant now... I did tinker some more with it, discovering I _can_ decrease redundancy in the profile as well as align it with the babel row-wise description... except KLE will helpfully transpose it and mess up my tedious effort... So will see whether I find any way forward, in that case of course I'd get in touch.

Yeah, I still think KLE probably a dead end... 

> Thanks for the `callum-oakley/keymap` reference! Indeed it seems a bit limited compared to KLE, yeah... we'll see slightly_smiling_face
 
I've been extending it slightly as I go, but not for other physical layouts yet.

manna-harbour commented:
> I redid the ANSI Lite layouts according to the comments and hope you agree they look way nicer now. As mentioned before, and even more so once the purpose of Lite variations was clarified, I think it is a really great entry level usage of Miryoku worth illustrating.

That looks great, thanks! I'll definitely be adding that! Actually, even your previous "complete" version was useful as it illustrated the other thumbs being mapped to the corresponding keys. But really this new version perfectly shows the intention of the lite mapping.

> In fact, I'll try again to get it going on my machines so I can still use mod tapping when my custom keyboard is not available. Thanks!

Great!

UncleCJ commented:
> > I redid the ANSI Lite layouts according to the comments and hope you agree they look way nicer now. As mentioned before, and even more so once the purpose of Lite variations was clarified, I think it is a really great entry level usage of Miryoku worth illustrating.
> 
> That looks great, thanks! I'll definitely be adding that! Actually, even your previous "complete" version was useful as it illustrated the other thumbs being mapped to the corresponding keys. But really this new version perfectly shows the intention of the lite mapping.

Hmm, yes… assuming you’re not in a too great hurry, perhaps I’ll try some other variation and try to fit it into the documentation wording, or do you wanna do that last bit yourself? Collaborating like this is quite a peculiar thing (though Git expert is a significant part of my profile), but I find it quite rewarding to figure out :-)

The latest version had only the Nav layer and mods, but as I picked the 60% template to modify, I realized one probably would want the Function layer, and perhaps even “num pad”, except the layer modifiers (if that is what we call the thumb keys) looked so awkward strewn all over a regular keyboard, and the KMonad config doesn’t exclude any layers out of the box (which newcomers would be expecting)…

Anyway, as I follow you responding to my comments in sequence, I’ll just interject that indeed both my Corne and KMonad seem to work excellently with Miryoku now, I’ve seen your `custom_config.h` directions etc now… further customization iterations should be significantly more consistent :-)

Side note, sharing the passion - my partner is getting into resin key caps casting so will be inheriting the Falba MX Corne while soldering a new acrylic case low profile Choc Corne for me <3 Let’s see how my monkeytype and Colemak Club exercises come along and whether I can inspire some household competition ;-) Take care and have lovely days!

manna-harbour commented:
> @manna-harbour hi! I hope you’re not feeling pestered by these comments or suggestions - I fidget with them for my own benefit, because it’s enjoyable and with a slight hope it may be useful for others.

Sorry if I'm sometimes slow to reply. I definitely appreciate your efforts! Feel free to create new posts in the Discussions (https://github.com/manna-harbour/miryoku/discussions/8) if there are issues you'd like to explore further individually, or send me a DM on discord (https://github.com/manna-harbour/manna-harbour/tree/main/contact) if you'd like to chat. But this is your PR, so do with it what you will!

> With that in mind I wanted to write some general comments and questions. Please pardon I won’t be cross referencing right now.
> 
> 1. Regarding adding images of Lite layouts, QWERTY etc and perhaps attempting a nicer color theme (I am trying out imitating "hysp") - the more I look, the more I realize any contributions need to fit into the existing documentation and usage of illustrations. For instance there are both `layers` and `mapping` folders and so far I have focused on the previous… and even though it’s the Lite mapping, I think it is more of a layer illustration… Also acting on any comments obviously pile up quickly - I did the un-ghosted lite layout and in general… do give me any feedback if you feel there is a chance I can effectively contribute, otherwise don’t fret it, I am fully aware of my… limited awareness :-D

Layers and mapping diagrams are in general pretty separate, they just both happen to be using KLE at the moment. Layer diagrams will be covered by https://github.com/manna-harbour/miryoku/issues/72. Mapping diagrams I don't see as being so important. I'll be reorganising the mappings soon (standardising between Miryoku QMK and Miryoku ZMK and probably also combining the docs) and just the text descriptions should cover the majority of mappings. Although no doubt the diagrams are still helpful in many cases. But at the moment they are only provided where a KLE is already available for that keyboard.

The lite mapping is a special case though as it's a hybrid mapping, having only Base and partial Nav, and including the rest of the default keymap, and definitely benefits from your inclusion of the legends. But I think it's ok to keep it in with the other mappings for now, at least until Miryoku SVG can handle its generation.

> 2. It’s an obvious limitation not to have a continuous generation of the KLE diagrams… I noted some potential errors such as broken logo urls, lack of `AltGr` modifiers and the tab glyph may be funky… I may submit some trivial merge requests on things like that, assuming they are of any use

The logo URLs I just update when I otherwise need to change something. AltGr was left out deliberately but is included now with Miryoku SVG. I haven't notice the issue with tab though? From now, changes to the layer diagrams should be done via Miryoku SVG as that will soon replace the current KLE versions (well, the per layer diagrams at least, to begin).

> 3. Contingent on the continued use of KLE… My merge request on transposing the JSON makes sense to me

I think Miryoku SVG should be able to handle substituting legends into the current JSON as is, without needing any rearrangement, as it's done via `cpp` macros in the same way as Miryoku KMonad (and slightly less directly, Miryoku QMK and Miryoku ZMK). The python data structure being generated now has the complete left side before the right, for example, so there's no requirement for it to be row wise. But if the whole JSON was to be generated from scratch directly by Miryoku Babel, then transposing might make things simpler.

> the SVG render you mentioned feels too limited in comparison, though the most future proof path certainly sounds like visualizing ergogen… I saw some Miryoku QMK report (?), but one wants to stay on the “source” side rather than the generated result? Do let me know if there’s anything regarding the new customization documentation, babel or similar which you think may be promising, and I’d be happy to pivot my efforts. Continuing on KLE a bit, regarding the “transposing“, as I was reading about Kyria and other custom keyboards, it occurred that the thumb clusters logically are columns despite being quite horizontal. I’ve taken to your language of locating keys per finger, and in that sense, KLE being _row_ oriented become a severe design deficiency which makes no sense. Most, if not all of your space in babel and implementations in QMK and KMonad are also row-oriented though… though I expect ergogen to be flexible. Do you have any thoughts on that?

I'll have to get back to you on that one!

> 4. Minor thoughts… but my Crkbd has 1.5U central thumb keys and it felt a bit awkward to consider the middle ones “primary” and previously they had been Space bar and Enter, respectively. I can’t manage to author a macro for it yet but have a personal mod which swap primary and secondary, and then swap back Enter and Backspace. Do you have a strong opinion whether I may regret that, or do you even feel like adopting it?

So the comfort order of the thumb keys is the one assumption of the physical layout. Generally if you use a few boards I find you can be pretty flexible with the absolute positions. Ideally though the physical layout would be a perfect match for you hand. I don't recommend changing the relative positions though unless it's completely unavoidable (e.g. le chiffre) as it breaks your muscle memory compatibility with the ideal physical layout as well as almost all other boards. 

For crkbd, the 1.5u key is mostly just historical baggage and doesn't indicate that it is supposed to be the home thumb. The crkbd thumbs are shifted around 0.5u from their ideal positions though, so the 1.5u key is closer to the resting position than it should be, but it still shouldn't take long to get used to homing on the middle.

> Furthermore the apostrophe and semicolon swap tripped me up (Club Colemak emulation broke

Yeah the alternative alpha arrangements are really just the alphas. But there needs to be some accommodation for remapping on the host based on standard QWERTY. See https://github.com/manna-harbour/miryoku/issues/35. Perhaps by Extra having the swap and also being connected to an alternative swapped copy of Num / Sym...

> with the positive side effect that I rebased on your development branch and got ColemakDH extras working beautifully instead), 

Great!

> I’m considering expanding/altering QWERTY with the Nordic layout (as I may convert also my partner to the crkbd team…).

Nice! Yeah, better international support is something else I'd like to include... One approach to support localised layouts on the host would be to load the keycode translations into Miryoku Babel and have it output the localised codes. There are also plenty of options when using more generic international layouts on the host.

> 5. Almost forgot, what do you think of “autoshift“ keys? I read of someone going all the way down to 30 keys, and it seemed to make a lot of sense to have several of the symbols autoshifted as on the number row (though mapped like a num pad)

Auto shift is enabled in Miryoku QMK for numbers and symbols.  For Miryoku ZMK and Miryoku KMonad see https://github.com/manna-harbour/miryoku/issues/36.

> All in all, I even got KMonad working decently now and with each step appreciate the logic of Miryoku, so would be quite happy both to “paint within its lines” and to contribute to it in any way I can. Thank you so much!

Excellent!

For Miryoku SVG, as soon as the core functions are there I'd be happy to add you to the private repo if you'd like to play around with it. There are plenty of open issues you might be interested in. The colour scheme and other aspects of the styles, legends, support for other physical layouts (e.g. via ergogen), KLE JSON legend substitution, and of course any other issues you've noticed with the current diagrams.
